File: apps/mobile/src/services/SubscriptionService.ts
/**
 * SubscriptionService — Singleton RevenueCat wrapper.
 *
 * Uses RevenueCat's own anonymous ID (not Supabase) so subscriptions survive
 * app reinstalls. On init, restorePurchases() syncs any active App Store
 * subscription to the current device, covering the reinstall edge case.
 */

import { Platform } from 'react-native';
import Purchases, {
  type CustomerInfo,
  type PurchasesConfiguration,
  LOG_LEVEL,
} from 'react-native-purchases';
import { ENV } from '../config/env';
import { AppsFlyerService } from './AppsFlyerService';
import { Analytics } from './AnalyticsService';

const ENTITLEMENT_ID = 'Inner_Circle';

let initialized = false;

async function initialize(): Promise<boolean> {
  if (initialized) return true;

  try {
    const config: PurchasesConfiguration = {
      apiKey: ENV.REVENUECAT_IOS_API_KEY,
    };

    if (__DEV__) {
      Purchases.setLogLevel(LOG_LEVEL.DEBUG);
    }

    await Purchases.configure(config);
    initialized = true;

    // Send device identifiers (IDFA/IDFV) to RevenueCat for attribution
    try {
      Purchases.collectDeviceIdentifiers();
    } catch (e: any) {
      console.warn('[SubscriptionService] collectDeviceIdentifiers failed (non-fatal):', e);
      Analytics.track('subscription_collect_identifiers_failed', {
        error_type: 'collect_device_identifiers',
        error_code: e?.code,
        error_message: e?.message,
      });
    }

    // Send AppsFlyer ID to RevenueCat for S2S integration
    try {
      const afUID = await AppsFlyerService.getAppsFlyerUID();
      if (afUID) {
        Purchases.setAppsflyerID(afUID);
        console.log('[SubscriptionService] AppsFlyer ID set on RevenueCat:', afUID);
      }
    } catch (e: any) {
      console.warn('[SubscriptionService] setAppsflyerID failed (non-fatal):', e);
      Analytics.track('subscription_appsflyer_id_failed', {
        error_type: 'set_appsflyer_id',
        error_code: e?.code,
        error_message: e?.message,
      });
    }

    // Restore purchases to cover reinstall / new anonymous-ID scenarios.
    try {
      await Purchases.restorePurchases();
    } catch (restoreErr: any) {
      console.warn('[SubscriptionService] restorePurchases failed (non-fatal):', restoreErr);
      Analytics.captureException(restoreErr, {
        error_type: 'restore_purchases_init',
        error_code: restoreErr?.code,
      });
      Analytics.track('subscription_restore_init_failed', {
        error_type: 'restore_purchases_init',
        error_code: restoreErr?.code,
        error_message: restoreErr?.message,
      });
    }

    console.log('[SubscriptionService] Initialized');
    return true;
  } catch (error) {
    console.warn('[SubscriptionService] Init failed:', error);
    initialized = true;
    return false;
  }
}

function hasEntitlement(info: CustomerInfo | null): boolean {
  return !!info && ENTITLEMENT_ID in (info.entitlements.active ?? {});
}

async function checkSubscription(): Promise<boolean> {
  try {
    const info = await Purchases.getCustomerInfo();
    return hasEntitlement(info);
  } catch {
    return false;
  }
}

/**
 * Identify an authenticated user in RevenueCat.
 * Transfers any anonymous subscription to the authenticated user ID.
 */
async function logIn(userId: string): Promise<boolean> {
  try {
    const { customerInfo } = await Purchases.logIn(userId);
    return hasEntitlement(customerInfo);
  } catch (e: any) {
    console.warn('[SubscriptionService] logIn failed:', e);
    Analytics.captureException(e, {
      error_type: 'revenuecat_login',
      error_code: e?.code,
    });
    Analytics.track('subscription_login_rn_failed', {
      error_type: 'revenuecat_login',
      error_code: e?.code,
      error_message: e?.message,
    });
    return false;
  }
}

/**
 * Reset RevenueCat to a new anonymous user.
 * Call on sign-out so the next session gets a fresh customer ID.
 */
async function logOut(): Promise<void> {
  try {
    await Purchases.logOut();
  } catch (e: any) {
    console.warn('[SubscriptionService] logOut failed:', e);
    Analytics.track('subscription_logout_rn_failed', {
      error_type: 'revenuecat_logout',
      error_code: e?.code,
      error_message: e?.message,
    });
  }
}

async function restore(): Promise<boolean> {
  try {
    const info = await Purchases.restorePurchases();
    return hasEntitlement(info);
  } catch (e: any) {
    Analytics.captureException(e, {
      error_type: 'restore_purchases_manual',
      error_code: e?.code,
    });
    Analytics.track('subscription_restore_manual_rn_failed', {
      error_type: 'restore_purchases_manual',
      error_code: e?.code,
      error_message: e?.message,
    });
    return false;
  }
}

function addListener(
  callback: (info: CustomerInfo) => void,
): () => void {
  Purchases.addCustomerInfoUpdateListener(callback);
  return () => {};
}

function isInitialized(): boolean {
  return initialized;
}

export const SubscriptionService = {
  initialize,
  logIn,
  logOut,
  checkSubscription,
  restore,
  addListener,
  hasEntitlement,
  isInitialized,
  ENTITLEMENT_ID,
};

File: apps/ios-swift/LockedIn/Features/Subscription/SubscriptionService.swift
//
//  SubscriptionService.swift
//  LockedIn
//
//  Ported from `apps/mobile/src/services/SubscriptionService.ts`.
//
//  RevenueCat wrapper. Uses RevenueCat's own anonymous ID (not Supabase) so
//  subscriptions survive app reinstalls. On `configure(apiKey:)`, we eagerly
//  call `restorePurchases()` to sync any active App Store subscription to the
//  current device, covering the reinstall edge case.
//
//  Entitlement key: `Inner_Circle` (do not change — locked by Agent A audit).
//

import Foundation
import RevenueCat
import AppsFlyerLib

/// Singleton wrapper around the RevenueCat SDK.
///
/// The RN code's `SubscriptionService` is an object literal; in Swift we
/// expose this as `SubscriptionService.shared`. All async methods bridge to
/// RevenueCat's callback API via Swift Concurrency.
@MainActor
public final class SubscriptionService {

    // MARK: - Public constants

    /// RevenueCat entitlement identifier that gates `isSubscribed`.
    /// Confirmed by Agent A backend audit. Do NOT change.
    public static let entitlementID = "Inner_Circle"

    // MARK: - Singleton

    public static let shared = SubscriptionService()

    private init() {}

    // MARK: - State

    private var isConfigured = false

    /// `true` once `configure(apiKey:)` has been called (success or failure).
    public var isInitialized: Bool { isConfigured }

    // MARK: - Configuration

    /// Configure the RevenueCat SDK. Idempotent — safe to call repeatedly.
    ///
    /// Mirrors RN `SubscriptionService.initialize()`:
    ///   1. `Purchases.configure(withAPIKey:)`
    ///   2. `Purchases.shared.collectDeviceIdentifiers()` for IDFA/IDFV
    ///   3. set AppsFlyer ID for S2S attribution (wired by coordinator)
    ///   4. `restorePurchases()` to cover reinstalls
    ///
    /// - Parameter apiKey: RevenueCat iOS public API key
    ///   (`REVENUECAT_IOS_API_KEY` xcconfig entry).
    @discardableResult
    public func configure(apiKey: String) async -> Bool {
        if isConfigured { return true }

        #if DEBUG
        Purchases.logLevel = .debug
        #else
        Purchases.logLevel = .info
        #endif

        // Note: `LockedInApp.configureSDKs()` already calls
        // `Purchases.configure(withAPIKey:)` at app startup. This call is
        // idempotent under the RevenueCat SDK — calling again with the same
        // key is a no-op. We re-call here to make the service self-contained
        // (the RN service did its own `Purchases.configure`).
        Purchases.configure(withAPIKey: apiKey)
        isConfigured = true

        // Send device identifiers (IDFA/IDFV) to RevenueCat for attribution.
        Purchases.shared.attribution.collectDeviceIdentifiers()

        // Hook AppsFlyer UID into RevenueCat for S2S attribution. RN does
        // this in SubscriptionService.ts:46. The AppsFlyer SDK exposes the
        // UID synchronously after `start()` runs in the AppDelegate.
        let uid = AppsFlyerLib.shared().getAppsFlyerUID()
        if !uid.isEmpty {
            Purchases.shared.attribution.setAppsflyerID(uid)
        }

        // Restore purchases on launch to cover reinstall / new anonymous-ID
        // scenarios. Best-effort; failure is logged + captured.
        do {
            _ = try await Purchases.shared.restorePurchases()
        } catch {
            let nsError = error as NSError
            AnalyticsService.shared.captureException(error, properties: [
                "context": "subscription_restore_init",
                "revenuecat_error_code": nsError.code,
                "revenuecat_error_domain": nsError.domain,
            ])
            AnalyticsService.shared.track("subscription_restore_failed", properties: [
                "revenuecat_error_code": nsError.code,
                "revenuecat_error_domain": nsError.domain,
                "error_message": error.localizedDescription,
            ])
        }

        return true
    }

    // MARK: - Identity

    /// Identify an authenticated user in RevenueCat. Transfers any anonymous
    /// subscription to the authenticated user ID.
    ///
    /// - Parameter uuid: The **Supabase `auth.users.id` UUID** (NOT email,
    ///   NOT RevenueCat ID). This is the link between Supabase user identity
    ///   and RevenueCat.
    /// - Returns: `true` if the resulting customer info has the
    ///   `Inner_Circle` entitlement active.
    @discardableResult
    public func logIn(uuid: String) async -> Bool {
        do {
            let result = try await Purchases.shared.logIn(uuid)
            return Self.hasEntitlement(in: result.customerInfo)
        } catch {
            let nsError = error as NSError
            AnalyticsService.shared.captureException(error, properties: [
                "context": "subscription_login",
                "revenuecat_error_code": nsError.code,
                "revenuecat_error_domain": nsError.domain,
            ])
            AnalyticsService.shared.track("subscription_login_failed", properties: [
                "revenuecat_error_code": nsError.code,
                "revenuecat_error_domain": nsError.domain,
                "error_message": error.localizedDescription,
            ])
            return false
        }
    }

    /// Reset RevenueCat to a new anonymous user. Call on sign-out so the next
    /// session gets a fresh customer ID.
    public func logOut() async {
        do {
            _ = try await Purchases.shared.logOut()
        } catch {
            // P2 — error event only, no exception capture per spec.
            let nsError = error as NSError
            AnalyticsService.shared.track("subscription_logout_failed", properties: [
                "revenuecat_error_code": nsError.code,
                "revenuecat_error_domain": nsError.domain,
                "error_message": error.localizedDescription,
            ])
        }
    }

    // MARK: - Entitlement checks

    /// Returns `true` if the entitlement `Inner_Circle` is active on the given
    /// customer info object.
    public static func hasEntitlement(in info: CustomerInfo?) -> Bool {
        guard let info else { return false }
        return info.entitlements.active[entitlementID] != nil
    }

    /// Convenience: fetch latest customer info and report whether the user has
    /// the `Inner_Circle` entitlement. Mirrors RN `checkSubscription()`.
    public func checkEntitlements() async -> Bool {
        do {
            let info = try await Purchases.shared.customerInfo()
            return Self.hasEntitlement(in: info)
        } catch {
            let nsError = error as NSError
            AnalyticsService.shared.captureException(error, properties: [
                "context": "subscription_entitlement_check",
                "revenuecat_error_code": nsError.code,
                "revenuecat_error_domain": nsError.domain,
            ])
            AnalyticsService.shared.track("subscription_entitlement_check_failed", properties: [
                "revenuecat_error_code": nsError.code,
                "revenuecat_error_domain": nsError.domain,
                "error_message": error.localizedDescription,
            ])
            return false
        }
    }

    /// Fetch the current `CustomerInfo` from RevenueCat. Returns `nil` on error.
    public func customerInfo() async -> CustomerInfo? {
        try? await Purchases.shared.customerInfo()
    }

    // MARK: - Offerings

    /// Fetch RevenueCat offerings (packages). Returns `nil` on error.
    public func offerings() async -> Offerings? {
        try? await Purchases.shared.offerings()
    }

    // MARK: - Purchases

    /// Distinguishes a user-cancelled purchase (no error UI) from a genuine
    /// failure (surface an error) and from success.
    public enum PurchaseOutcome: Equatable { case subscribed, cancelled, failed }

    /// Purchase the given package. Returns `true` if the resulting customer
    /// info has the `Inner_Circle` entitlement active.
    @discardableResult
    public func purchase(package: Package) async -> Bool {
        await purchaseOutcome(package: package) == .subscribed
    }

    /// Purchase the given package, reporting whether the user cancelled vs a
    /// real error occurred — used by the custom paywall to avoid flashing an
    /// error banner when the user simply backs out of the App Store sheet.
    @discardableResult
    public func purchaseOutcome(package: Package) async -> PurchaseOutcome {
        do {
            let result = try await Purchases.shared.purchase(package: package)
            if result.userCancelled { return .cancelled }
            return Self.hasEntitlement(in: result.customerInfo) ? .subscribed : .failed
        } catch {
            let nsError = error as NSError
            AnalyticsService.shared.captureException(error, properties: [
                "context": "subscription_purchase",
                "revenuecat_error_code": nsError.code,
                "revenuecat_error_domain": nsError.domain,
                "package_identifier": package.identifier,
                "product_identifier": package.storeProduct.productIdentifier,
            ])
            AnalyticsService.shared.track("subscription_purchase_failed", properties: [
                "revenuecat_error_code": nsError.code,
                "revenuecat_error_domain": nsError.domain,
                "error_message": error.localizedDescription,
                "package_identifier": package.identifier,
                "product_identifier": package.storeProduct.productIdentifier,
            ])
            return .failed
        }
    }

    /// Restore previous purchases. Returns `true` if the restored customer
    /// info has the `Inner_Circle` entitlement active. Mirrors RN
    /// `restore()`.
    @discardableResult
    public func restorePurchases() async -> Bool {
        do {
            let info = try await Purchases.shared.restorePurchases()
            return Self.hasEntitlement(in: info)
        } catch {
            let nsError = error as NSError
            AnalyticsService.shared.captureException(error, properties: [
                "context": "subscription_restore_manual",
                "revenuecat_error_code": nsError.code,
                "revenuecat_error_domain": nsError.domain,
            ])
            AnalyticsService.shared.track("subscription_restore_manual_failed", properties: [
                "revenuecat_error_code": nsError.code,
                "revenuecat_error_domain": nsError.domain,
                "error_message": error.localizedDescription,
            ])
            return false
        }
    }
}
